library/src/main/java/com/datatorrent/lib/join/AbstractJoinOperator.java

```diff
@@ -0,0 +1,343 @@
+    // If the tuple is null means the join type is outer and return unmatched tuples from store.
+    Object value;
+    if (isLeft) {
+      if (tuple != null) {
+        value = store[1].getValidTuples(tuple);
+      } else {
+        value = store[1].getUnMatchedTuples();
+      }
+    } else {
+      if (tuple != null) {
+        value = store[0].getValidTuples(tuple);
+      } else {
+        value = store[0].getUnMatchedTuples();
+      }
+    }
+    // Join the input tuple with the joined tuples
+    if (value != null) {
+      ArrayList<TimeEvent> joinedValues = (ArrayList<TimeEvent>)value;
+      List<T> result = new ArrayList<>();
+      for (TimeEvent joinedValue : joinedValues) {
+        T output = createOutputTuple();
+        Object tupleValue = null;
+        if (tuple != null) {
+          tupleValue = tuple.getValue();
+        }
+        copyValue(output, tupleValue, isLeft);
+        copyValue(output, joinedValue.getValue(), !isLeft);
+        result.add(output);
+        joinedValue.setMatch(true);
+      }
+      if (tuple != null) {
+        tuple.setMatch(true);
+      }
+      if (result.size() != 0) {
+        outputPort.emit(result);
+      }
+    }
+  }
+
+  // Emit the unmatched tuples, if the strategy is outer join
+  @Override
+  public void endWindow()
+  {
+    if (strategy.equals(JoinStrategy.LEFT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN)) {
+      join(null, false);
+    }
+    if (strategy.equals(JoinStrategy.RIGHT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN)) {
+      join(null, true);
+    }
+    store[0].endWindow();
+    store[1].endWindow();
+  }
+
+  @Override
+  public void checkpointed(long windowId)
+  {
+    store[0].checkpointed(windowId);
+    store[1].checkpointed(windowId);
+  }
+
+  @Override
+  public void committed(long windowId)
+  {
+    store[0].committed(windowId);
+    store[1].committed(windowId);
+  }
+
+  /**
+   * Convert the given tuple to event
+   *
+   * @param tuple Given tuple to convert into event
+   * @return event
+   */
+  protected TimeEvent createEvent(Object tuple)
+  {
+    int idx = 0;
+    if (!isLeft) {
+      idx = 1;
+    }
+    if (timeFields != null) {
+      return new TimeEventImpl(getKeyValue(keys[idx], tuple), (Long)getTime(timeFields[idx], tuple), tuple);
+    } else {
+      return new TimeEventImpl(getKeyValue(keys[idx], tuple), Calendar.getInstance().getTimeInMillis(), tuple);
+    }
+  }
+
+  public void populateKeyFields()
+  {
+    this.keys = keyFieldStr.split(",");
+  }
+
+  public JoinStrategy getStrategy()
+  {
+    return strategy;
+  }
+
+  public void setStrategy(JoinStrategy strategy)
+  {
+    this.strategy = strategy;
+  }
+
+  public void setLeftStore(BackupStore lStore)
+  {
+    store[0] = lStore;
+  }
+
+  public void setRightStore(BackupStore rStore)
+  {
+    store[1] = rStore;
+  }
+
+  public void setKeyFields(String keyFieldStr)
+  {
+    this.keyFieldStr = keyFieldStr;
```

Comment:
Made these key, includeFields, timeFields  arrays as transient. So, I am not incorporating this review comment.